File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.cc
#include <iostream>
using namespace std;


void Move(int cnt, int start, int sub, int target)
{
	if (cnt == 1)
	{
		cout << start << " " << target << "\n";
	}
	if (cnt > 1)
	{
		Move(cnt - 1, start, target, sub);
		cout << start << " " << target << "\n";
		Move(cnt - 1, sub, start, target);
	}
	
}


int main()
{
	int input;
	cin >> input;
	int cnt = 1;
	for (int i = 0; i < input; i++)
	{
		cnt *= 2;
	}
	cnt--;
	cout << cnt << "\n";
	Move(input, 1, 2, 3);


	return 0;
}

File: 백준/Gold/11729. 하노이 탑 이동 순서/하노이 탑 이동 순서.cs
using System.Text;

class Program
{
    static StringBuilder stringBuilder = new StringBuilder();

    static void HanoiTower(int count, int start, int sub, int target)
    {
        if(count > 1)
        {
            HanoiTower(count - 1, start, target, sub);
            stringBuilder.Append($"{start} {target}\n");
            HanoiTower(count - 1, sub, start, target);
        }
        if(count == 1)
        {
            stringBuilder.Append($"{start} {target}\n");
        }
    }

    static void Main()
    {
        int input = int.Parse(Console.ReadLine());

        Console.WriteLine(Math.Pow(2, input) - 1);

        HanoiTower(input, 1, 2, 3);

        Console.WriteLine(stringBuilder.ToString());

    }
}
